### PR TITLE
Allow API server to more easily run in an HA environment

### PR DESCRIPTION
**Pull request by smarterclayton**

### CODE REVIEW

File: pkg/cmd/server/api/types.go [hunk 1]
```diff
@@ -533,7 +533,12 @@ type KubernetesMasterConfig struct {
 	APILevels []string
 	// MasterIP is the public IP address of kubernetes stuff.  If empty, the first result from net.InterfaceAddrs will be used.
 	MasterIP string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: before we introduce parallel variables for things we might want to load from envvars, we should decide if we want parallel fields for everything we might want to do that for. Things that come to mind:

1. HA-specific fields, like IPs/hostnames
2. Secret fields, like `clientSecret` for OAuth identity providers, or `bindPassword` for the upcoming LDAP identity provider

I would sort of like to let those fields take direct values, or envvars, or values from a conf file, or some other yet undetermined mechanism (etcd lookup, etc). @deads2k is opposed to letting values contain a syntax like `${ENVVAR}`, `${FILE:/path/to/conf/file:KEY_NAME}`. I sort of like the unification, but I can see pitfalls. Implications (good, bad, and neutral):
* you can only do it for string fields (or yaml unmarshalling won't preserve the value)
* a single field is sufficient for any number of resolution sources
* you have to be careful not to resolve values when reading a config in order to mutate and write it back out (we don't do this yet, but we may someday)
* you have to be careful to resolve values before using/validating the config

Comment (smarterclayton) [inline pkg/cmd/server/api/types.go]: It's possible anything secret should be a file. Another alternative is to
allow a deterministic layering by converting a subset of env var names
directly to the config fields (KUBERNETES_MASTER_CONFIG_MASTER_IP = ...).
I don't like doing variable expansion.

I do like being explicit about the config source, but practically the
reason to use env is to make us easy to run in a container, and that should
be our guiding use case (does this make running in a container easy and
does it reduce complexity in that env).

On Jul 7, 2015, at 12:06 AM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/cmd/server/api/types.go
<https://github.com/openshift/origin/pull/3599#discussion_r34006483>:

> @@ -533,7 +533,12 @@ type KubernetesMasterConfig struct {
>  	APILevels []string
>  	// MasterIP is the public IP address of kubernetes stuff.  If empty, the first result from net.InterfaceAddrs will be used.
>  	MasterIP string

before we introduce parallel variables for things we might want to load
from envvars, we should decide if we want parallel fields for everything we
might want to do that for. Things that come to mind:

   1. HA-specific fields, like IPs/hostnames
   2. Secret fields, like clientSecret for OAuth identity providers, or
   bindPassword for the upcoming LDAP identity provider

I would sort of like to let those fields take direct values, or envvars, or
values from a conf file, or some other yet undetermined mechanism (etcd
lookup, etc). @deads2k <https://github.com/deads2k> is opposed to letting
values contain a syntax like ${ENVVAR}, ${FILE:/path/to/conf/file:KEY_NAME}.
I sort of like the unification, but I can see pitfalls. Implications (good,
bad, and neutral):

   - you can only do it for string fields (or yaml unmarshalling won't
   preserve the value)
   - a single field is sufficient for any number of resolution sources
   - you have to be careful not to resolve values when reading a config in
   order to mutate and write it back out (we don't do this yet, but we may
   someday)
   - you have to be careful to resolve values before using/validating the
   config

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3599/files#r34006483>.

**Comment by danmcp** [comment]
@smarterclayton Bump

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
This needs review, last piece blocking HA of masters.

**Comment by smarterclayton** [comment]
I need to let listen be overridable for master and master api, so that you can run two masters on different machines with different ips with one config

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**